File: aat-android/src/main/java/ch/bailu/aat/services/cache/CacheService.java
package ch.bailu.aat.services.cache;

import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.dispatcher.AppBroadcaster;
import ch.bailu.aat_lib.dispatcher.BroadcastReceiver;
import ch.bailu.aat_lib.preferences.OnPreferencesChanged;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.preferences.system.SolidCacheSize;
import ch.bailu.aat_lib.service.VirtualService;
import ch.bailu.aat_lib.service.cache.CacheServiceInterface;
import ch.bailu.aat_lib.service.cache.Obj;
import ch.bailu.aat_lib.service.cache.ObjBroadcastReceiver;
import ch.bailu.aat_lib.util.MemSize;
import ch.bailu.aat_lib.util.WithStatusText;


public final class CacheService extends VirtualService implements CacheServiceInterface, OnPreferencesChanged, WithStatusText {



    public final ObjectTable table=new ObjectTable();
    public final ObjectBroadcaster broadcaster;

    public final AppContext appContext;

    private final SolidCacheSize slimit;

    public CacheService(AppContext sc) {
        appContext = sc;
        broadcaster = new ObjectBroadcaster(sc);

        slimit = new SolidCacheSize(sc.getStorage());
        slimit.register(this);

        table.limit(this, slimit.getValueAsLong());

        sc.getBroadcaster().register(onFileProcessed, AppBroadcaster.FILE_CHANGED_INCACHE);
    }

    public void onLowMemory() {
        table.limit(this, MemSize.MB);
        slimit.setIndex(1);
    }

    public Obj getObject(String id, Obj.Factory factory) {

        return table.getHandle(id, factory, this);
    }

    public Obj getObject(String id) {
        return table.getHandle(id, appContext);
    }

    @Override
    public void appendStatusText(StringBuilder builder) {
        table.appendStatusText(builder);
    }


    public void close() {
        appContext.getBroadcaster().unregister(onFileProcessed);

        slimit.unregister(this);
        broadcaster.close();
        table.close(this);
    }


    public void addToBroadcaster(ObjBroadcastReceiver b) {
        broadcaster.put(b);
    }

    private final BroadcastReceiver onFileProcessed = new BroadcastReceiver() {
        @Override
        public void onReceive(Object... objs) {
            table.onObjectChanged(CacheService.this, objs);
        }
    };

    @Override
    public void onPreferencesChanged(StorageInterface s, String key) {
        if (slimit.hasKey(key)) {
            table.limit(this, slimit.getValueAsLong());
        }
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/cache/CacheService.kt
package ch.bailu.aat_lib.service.cache

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.broadcaster.BroadcastReceiver
import ch.bailu.aat_lib.preferences.OnPreferencesChanged
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.preferences.system.SolidCacheSize
import ch.bailu.aat_lib.service.VirtualService
import ch.bailu.aat_lib.util.MemSize
import ch.bailu.aat_lib.util.WithStatusText

class CacheService(val appContext: AppContext) : VirtualService(), CacheServiceInterface,
    OnPreferencesChanged, WithStatusText {

    private val table: ObjectTable = ObjectTable()
    private val broadcaster: ObjectBroadcaster = ObjectBroadcaster(appContext)

    private val solidCacheSize = SolidCacheSize(appContext.storage)

    private val onFileProcessed: BroadcastReceiver = BroadcastReceiver{ args ->
        table.onObjectChanged(this@CacheService, *args)
    }

    init {
        solidCacheSize.register(this)
        table.limit(this, solidCacheSize.valueAsLong)
        appContext.broadcaster.register(AppBroadcaster.FILE_CHANGED_INCACHE, onFileProcessed)
    }

    override fun onLowMemory() {
        table.limit(this, MemSize.MB)
        solidCacheSize.index = 1
    }

    override fun getObject(path: String, factory: Obj.Factory): Obj {
        return table.getHandle(path, factory, this)
    }

    override fun getObject(id: String): Obj {
        return table.getHandle(id)
    }

    override fun appendStatusText(builder: StringBuilder) {
        table.appendStatusText(builder)
    }

    override fun close() {
        appContext.broadcaster.unregister(onFileProcessed)

        solidCacheSize.unregister(this)
        broadcaster.close()
        table.close(this)
    }

    override fun addToBroadcaster(obj: ObjBroadcastReceiver) {
        broadcaster.put(obj)
    }

    override fun onPreferencesChanged(storage: StorageInterface, key: String) {
        if (solidCacheSize.hasKey(key)) {
            table.limit(this, solidCacheSize.valueAsLong)
        }
    }
}
